[Skia] Integrate building Skia with CMake

aperezdc opened the pull request:
#### 3b54e45d63b84f8ab8d70179184c45be78386622
<pre>
[Skia] Integrate building Skia with CMake
<a href="https://bugs.webkit.org/show_bug.cgi?id=269006">https://bugs.webkit.org/show_bug.cgi?id=269006</a>

Reviewed by Carlos Garcia Campos.

Add CMake build definitions to compile Skia. This compiles only those
parts of Skia that will be needed at first by the WPE port, and while
it was written mostly by hand with the source file lists and some
inspiration from the Skia GN-based build system. The only missing bit
is a call to add_directory() to wire the subproject into the rest of
the CMake build, which will be done in a follow-up.

Additionally, this also includes a implementation of Skia&apos;s
GrGLMakeNativeInterface() that calls into libepoxy to resolve GL/EGL
entry points. This ensures that the functions used by Skia are the same
as the rest of WebKit and that libWPEWebKit is only ever linked at build
time against libepoxy and not concrete GL/EGL libraries.

* Source/ThirdParty/skia/CMakeLists.txt: Added.
* Source/ThirdParty/skia/README.WebKit: Added.
* Source/ThirdParty/skia/ext-skia-configure: Added.
* Source/ThirdParty/skia/include/gpu/gl/epoxy/GrGLMakeEpoxyEGLInterface.h: Added.
* Source/ThirdParty/skia/src/gpu/ganesh/gl/epoxy/GrGLMakeEpoxyEGLInterface.cpp: Added.
(epoxy_get_gl_proc):
(GrGLMakeEpoxyEGLInterface):
* Source/ThirdParty/skia/src/gpu/ganesh/gl/epoxy/GrGLMakeNativeInterface_epoxy.cpp: Added.
(GrGLMakeNativeInterface):
* Source/cmake/OptionsWPE.cmake: Introduce an USE_CAIRO build option,
set to TRUE for now, and make checking for Cairo, Fontconfig, and
Freetype guarded by it. Once we add the Skia variant it will not be needed
to list these directly because the Skia font manager will be used.

Canonical link: <a href="https://commits.webkit.org/274359@main">https://commits.webkit.org/274359@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/24b56d0e3bc20cf7a4dda3155c88683a961bf150

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/38882 "Failed to checkout and rebase branch from PR 24100") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/48/builds/17813 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/14/builds/41227 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/41413 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/34535 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/49/builds/20659 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/51/builds/15162 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/41413 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [❌ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/39455 "Failed to checkout and rebase branch from PR 24100") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/49/builds/20659 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/14/builds/41227 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/41413 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/49/builds/20659 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/14/builds/41227 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/42690 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/32328 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/49/builds/20659 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/14/builds/41227 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/42690 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc-arm64~~](https://ews-build.webkit.org/#/builders/12/builds/38507 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/13692 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/11288 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/42690 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/15300 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/14/builds/41227 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/35/builds/45514 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/14965 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/35/builds/45514 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5065 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/14785 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

donny-dont commented on Source/ThirdParty/skia/CMakeLists.txt:
```diff
@@ -0,0 +1,1019 @@
+if (SKIA_ROOT_DIR OR NOT "$ENV{SKIA_ROOT_DIR}" STREQUAL "")
+    if (NOT SKIA_ROOT_DIR)
+        set(SKIA_ROOT_DIR "$ENV{SKIA_ROOT_DIR")
+    endif ()
```
This part old? Looks like it was supporting a `SKIA_ROOT_DIR` out of source.

aperezdc replied:
We were debating whether to remove this or not, without much of a conclusion. Personally I would prefer it gone; but others have found it useful during development.

@alexgcastro, @nikolaszimmermann and @carlosgcampos: Could we remove this, or is there still any good reason why we would need to keep it a bit longer?

carlosgcampos replied:
I would keep it for now. We are still seeing differences in some cases and it's very convenient to be able to try with external skia build to compare. We can remove it once we are sure we are building skia with the right options.

carlosgcampos replied:
We can add a FIXME explaining this is temporary and will be removed eventually.

alexgcastro replied:
Yeah, please keep it, I think it is very valuable even in the future to check if the compilation we are doing with CMake provides the same results and performance, it simplifies testing the original gn compilation and compare the results.

nikolaszimmermann replied:
Agreed. Let us keep it for now and add a FIXME.

aperezdc replied:
I'll add a comment about why we are keeping it for now, and also gate it behind `DEVELOPER_MODE` so that way it is clear that the intention is to have it as an aid for developers.

donny-dont commented on Source/ThirdParty/skia/CMakeLists.txt:
```diff
@@ -0,0 +1,1019 @@
+        set(SKIA_ROOT_DIR "$ENV{SKIA_ROOT_DIR")
+    endif ()
+
+    find_path(SKIA_INCLUDE_DIR
+        NAMES include/private/base/SkFeatures.h
+        HINTS ${SKIA_INCLUDE_DIR} ${SKIA_ROOT_DIR}
+        NO_DEFAULT_PATH
+        REQUIRED
+    )
+    find_library(SKIA_LIBRARY
+        NAMES skia libskia
+        HINTS ${SKIA_LIBRARY_PATH} ${SKIA_ROOT_DIR}/out ${SKIA_ROOT_DIR}/out/Static
+        NO_DEFAULT_PATH
+        REQUIRED
+    )
+    find_path(SKIA_NINJA_RULES
+        NAMES skia.ninja
+        HINTS ${SKIA_LIBRARY_PATH} ${SKIA_ROOT_DIR}
+        PATH_SUFFIXES out/obj out/Static/obj
+        NO_DEFAULT_PATH
+        REQUIRED
+    )
```
I don't think the finds are needed with the assumption that the tree is in source.

aperezdc replied:
Correct, let's see if we can remove all the support for external Skia builds.

aperezdc replied:
We would like to keep this for now, as discussed above.

donny-dont commented on Source/ThirdParty/skia/CMakeLists.txt:
```diff
@@ -0,0 +1,1019 @@
+    find_package(OpenGLES2 REQUIRED)
+    find_package(Freetype 2.9.0 REQUIRED)
+    find_package(Fontconfig 2.13.0 REQUIRED)
+    find_package(PNG REQUIRED)
```
These is already present in `Source/cmake/Options${PORT_NAME}.cmake` so they aren't needed. If a specific library version is needed it should go in there not here.

aperezdc replied:
I hope this just goes away if we remove the support for external Skia builds, too.

carlosgcampos replied:
I agree we don't want to find packages twice, because of our policy dependency we can't make skia depend on a different version, so it's ok to have a single find package in the main cmake file for shared dependencies. But I also agree with having find_package where they are actually needed, so what I would do here is:

```
find_package(EGL REQUIRED)
find_package(OpenGLES2 REQUIRED)
```

I don't think we need these for WPE and GTK since epoxy is our dependency, so I would just remove those for now.

```
find_package(Freetype 2.9.0 REQUIRED)
find_package(Fontconfig 2.13.0 REQUIRED)
```

Those are in the main file because WebCore depends on those, but that's not the case when building with skia, so I would move those inside a if USE_CAIRO in the main file and keep them here too.

```
find_package(PNG REQUIRED)
```

This is required by the image decoders that are used when building with skia, so I would remove this from here and just use the one in the main file.

carlosgcampos replied:
Oh, I realized now that this is for the external skia.

aperezdc replied:
I'm going to simplify both a bit with the suggestions, thanks to both!

aperezdc replied:
Done!

donny-dont commented on Source/ThirdParty/skia/CMakeLists.txt:
```diff
@@ -0,0 +1,1019 @@
+find_package(Epoxy 1.5.4 REQUIRED)
+find_package(Freetype 2.9.0 REQUIRED)
+find_package(Fontconfig 2.13.0 REQUIRED)
+find_package(PNG REQUIRED)
```
Again libraries already found so don't need to re-find them. Update the minimum version if necessary.

aperezdc replied:
This is a bit of a philosophical issue, as for example libpng is always needed regardless of which font/EGL/GL support libraries each port may want to use along with Skia, so I would argue at least the libpng one should be here, having the dependency right where it's used. OTOH we have the convention of writing down all the dependencies in the `Options<port>.cmake` files, but I feel that often the `find_package()` is “far” from where the dependency is used...

At any rate, let's go with the convention for now, I'll leave these be into `OptionsWPE.cmake` for now.

donny-dont replied:
I think the assumption should be that now that this is in `main` you're never going to build it outside WebKit and you should act accordingly. Its using some WebKit macros too.

If the required versions are tighter here than a normal WPE/GTK build then maybe the `Options${PORT}.cmake` should do something like

```
if (USE_SKIA AND FreeType_VERSION VERSION_LESS 2.9.0)
    message(FATAL_ERROR "FreeType version needs to be >= 2.9.0; Current version is ${FreeType_VERSION}")
endif ()
```

carlosgcampos replied:
So, my previous comment still applies, I would remove epoxy and PNG from here, keep the freetype and fontconfig, and move freetype and fontconfig inside a USE_CAIRO if in the main file.

aperezdc replied:
I think @carlosgcampos's proposal is a nice middle ground, I'll go with that.

aperezdc replied:
@donny-dont No, so far Skia doesn't require newer dependencies that versions we were already listing, so it's fine to pick the shared targets from `OptionsWPE.cmake` when they are already there 👌🏼

donny-dont commented on Source/ThirdParty/skia/CMakeLists.txt:
```diff
@@ -0,0 +1,1019 @@
+    SK_DISABLE_TRACING
+    SK_DISABLE_WEBGL_INTERFACE
+
+    SK_ENABLE_AVX512_OPTS
```
Not sure what these options are yet but I see some architecture checks below so should that go there?

aperezdc replied:
It's “opts” for “optimizations”; in this case it's safe to always set the macro because there are additional guards _also_ checking for x86 in the relevant source files. I guess it could be moved below in the `if (WTF_CPU_X86 OR WTF_CPU_X86_64)` block.

donny-dont commented on Source/ThirdParty/skia/CMakeLists.txt:
```diff
@@ -0,0 +1,1019 @@
+    src/gpu/Swizzle.cpp
+    src/gpu/TiledTextureUtils.cpp
+
+    src/gpu/ganesh/ClipStack.cpp
+    src/gpu/ganesh/Device.cpp
+    src/gpu/ganesh/Device_drawTexture.cpp
+    src/gpu/ganesh/GrAttachment.cpp
+    src/gpu/ganesh/GrAuditTrail.cpp
+    src/gpu/ganesh/GrBackendSemaphore.cpp
+    src/gpu/ganesh/GrBackendSurface.cpp
+    src/gpu/ganesh/GrBackendTextureImageGenerator.cpp
```
Do you want to move the ganesh stuff to its own variable and then when you add the other backend you could switch between the two for the sources?

aperezdc replied:
We are not completely decided on it, but for now we plan to always build Ganesh and add an option to build Graphite as well and choose at runtime among both. I understand you may want to use only Graphite for the PS and/or Win ports; so I think we can split the source list into variables when we get to that point. Does that sound good?

donny-dont replied:
Sure

donny-dont reviewed (approved):
Seems like you have some old code from when you all were using a Skia outside its new location of `Source/ThirdParty/skia`.

Overall looks fine. Not going to nit pick the portions that obviously need to be moved into a `Platform${PORT_NAME}.cmake` definition. Sure it'll get fleshed out more when I start up builds for the other ports.

carlosgcampos commented on Source/ThirdParty/skia/CMakeLists.txt:
```diff
@@ -0,0 +1,1029 @@
+# to ensure that behaviour and performance of the integrated build matches
+# Skia built with its own GN-based build system.
+#
+# TODO: This should be removed once it is no longer useful.
```
Use FIXME instead of TODO

aperezdc committed:
[Skia] Integrate building Skia with CMake
https://bugs.webkit.org/show_bug.cgi?id=269006

Reviewed by Carlos Garcia Campos.

Add CMake build definitions to compile Skia. This compiles only those
parts of Skia that will be needed at first by the WPE port, and while
it was written mostly by hand with the source file lists and some
inspiration from the Skia GN-based build system. The only missing bit
is a call to add_directory() to wire the subproject into the rest of
the CMake build, which will be done in a follow-up.

Additionally, this also includes a implementation of Skia's
GrGLMakeNativeInterface() that calls into libepoxy to resolve GL/EGL
entry points. This ensures that the functions used by Skia are the same
as the rest of WebKit and that libWPEWebKit is only ever linked at build
time against libepoxy and not concrete GL/EGL libraries.

* Source/ThirdParty/skia/CMakeLists.txt: Added.
* Source/ThirdParty/skia/README.WebKit: Added.
* Source/ThirdParty/skia/ext-skia-configure: Added.
* Source/ThirdParty/skia/include/gpu/gl/epoxy/GrGLMakeEpoxyEGLInterface.h: Added.
* Source/ThirdParty/skia/src/gpu/ganesh/gl/epoxy/GrGLMakeEpoxyEGLInterface.cpp: Added.
(epoxy_get_gl_proc):
(GrGLMakeEpoxyEGLInterface):
* Source/ThirdParty/skia/src/gpu/ganesh/gl/epoxy/GrGLMakeNativeInterface_epoxy.cpp: Added.
(GrGLMakeNativeInterface):
* Source/cmake/OptionsWPE.cmake: Introduce an USE_CAIRO build option,
set to TRUE for now, and make checking for Cairo, Fontconfig, and
Freetype guarded by it. Once we add the Skia variant it will not be needed
to list these directly because the Skia font manager will be used.

Canonical link: https://commits.webkit.org/274359@main
